functions/go/set-namespace/transformer/namespace.go

```diff
@@ -98,70 +70,208 @@ func (p *NamespaceTransformer) Config(o *fn.KubeObject) error {
+	if p.namespaceSelector != "" {
+		if len(nsCount) == 0 {
+			return p.namespaceSelector, true
+		}
+		p.Errors = append(p.Errors,
+			"found Namespace objects from the input resources, "+
+				"you cannot use `namespaceSelector` in FunctionConfig together with Namespace objects")
+		return "", false
+	}
+	if len(nsCount) > 1 {
+		msg := fmt.Sprintf("cannot accept more than one Namespace objects from the input resources, found %v",
+			nsCount)
+		p.Errors = append(p.Errors, msg)
+		return "", false
+	}
+	if len(nsCount) == 1 {
+		// Use the namespace object as the matching namespace if `namespaceSelector` is not given.
+		oldNs := reflect.ValueOf(nsCount).MapKeys()[0].String()
+		if nsCount[oldNs] > 1 {
+			msg := fmt.Sprintf("found more than one Namespace objects of the same name %v", oldNs)
+			p.Errors = append(p.Errors, msg)
+			return "", false
+		}
+		return oldNs, true
+	}
+	if len(fromResources) > 1 {
+		msg := fmt.Sprintf("all input namespace-scoped resources should be under the same namespace "+
+			"but found different namespaces: %v ", strings.Join(fromResources, ","))
+		p.Errors = append(p.Errors, msg)
+		return "", false
 	}
+	return fromResources[0], true
+}
+
+type NamespaceTransformer struct {
+	NewNamespace      string
+	namespaceSelector string
+	DependsOnMap      map[string]bool
+	Errors            []string
 }
 
-func (p *NamespaceTransformer) UpdateAnnotation(o *fn.KubeObject) {
-	anno, ok := o.GetAnnotations()[dependsOnAnnotation]
-	if !ok {
-		return
+func NewNamespace(obj *fn.KubeObject, path []string) *Namespace {
+	setter := func(newNs string) {
+		obj.Set(newNs, path...)
 	}
-	if !namespacedResourcePattern.MatchString(anno) {
-		return
+	annoSetter := func(newAnno string) {
+		obj.SetAnnotation(dependsOnAnnotation, newAnno)
 	}
-	segments := strings.Split(anno, "/")
-	dependsOnkey := dependsOnKeyPattern(segments[groupIdx], segments[kindIdx], segments[nameIdx])
-	if ok := p.DependsOnMap[dependsOnkey]; ok {
-		segments[namespaceIdx] = p.Namespace
-		newAnno := strings.Join(segments, "/")
-		o.SetAnnotation(dependsOnAnnotation, newAnno)
+	annoGetter := func() string {
+		group := obj.GetAPIVersion()
+		if i := strings.Index(obj.GetAPIVersion(), "/"); i > -1 {
+			group = group[:i]
+		}
+		return dependsOnKeyPattern(group, obj.GetKind(), obj.GetName())
 	}
+	return &Namespace{
+		Name:          obj.GetStringOrDie(path...),
+		IsNamespace:   obj.IsGVK("v1", "Namespace"),
+		DependsOnAnno: obj.GetAnnotations()[dependsOnAnnotation],
+		setter:        setter,
+		annoGetter:    annoGetter,
+		annoSetter:    annoSetter,
+	}
+}
+
+type Namespace struct {
+	Name          string
+	IsNamespace   bool
+	DependsOnAnno string
+	obj           *fn.KubeObject
+	setter        func(newNs string)
+	annoGetter    func() string
+	annoSetter    func(newDependsOnAnno string)
+}
+
+func (n *Namespace) Set(newNs string) {
+	n.setter(newNs)
+}
+
+func (n *Namespace) SetDependsOnAnno(newDependsOnAnno string) {
```

Comment:
Nit: I would just spell out `SetDependsOnAnnotation` etc rather than using the "Anno" abbreviation.

You _could_ argue that annotation is just an implementation detail, and make it something like `SetDependsOn(gvknn dependsOn)` -  that's consistent with e.g. a future method that gets all dependencies of an object.

Reply:
renamed. I won't argue but agree +50 on the full spell. That's exactly the buggy moment I got confused when first see the short form.